### ISSUE TITLE
HausdorffERLoss is not differentiable

### ISSUE THREAD

**Issue by shijianjian** [JunMa11/SegLoss#21]
I wrote a doctest for the erosion based hausdorff loss and I found it is not differentiable:
```
    Examples:
        >>> image_preds = torch.tensor([[
        ...    [0, 0.1, 0.1, 0],
        ...    [0, 0.3, 0.6, 0],
        ...    [0, 0.4, 0.8, 0],
        ...    [0, 0, 0, 0],
        ... ]], requires_grad=True)[None]
        >>> image_gt = torch.tensor([[
        ...    [0, 1, 0, 0],
        ...    [1, 1, 1, 0],
        ...    [1, 1, 1, 0],
        ...    [0, 0, 0, 0],
        ... ]])[None]
        >>> hdloss = HausdorffERLoss(from_logits=False)
        >>> hdloss(image_preds, image_gt)
        tensor(0.0625)
        >>> hdloss(image_preds, image_gt).backward()
```
I would suspect something is missing from your implementation.

**Comment by JunMa11** [comment]
Hi, @shijianjian 
I do not implement the erosion-based HD loss.
I only use the DTM-based HD loss in my repo.

**Comment by shijianjian** [comment]
Isn't this the erosion based HD loss?

https://github.com/JunMa11/SegLoss/blob/master/losses_pytorch/hausdorff.py#L83-L173

**Comment by JunMa11** [comment]
Oh, I apologize for the inconvenience.

The erosion-based HD loss was copied from https://github.com/SilmarilBearer/HausdorffLoss and I have not tested it.

https://github.com/JunMa11/SegLoss/blob/71b14900e91ea9405d9705c95b451fc819f24c70/losses_pytorch/hausdorff.py#L15

I would recommend raising an issue in the original repo.

**Comment by lucafei** [comment]
is the HausdorffDLoss possible for multi class segmentation？

**Comment by shijianjian** [comment]
@JunMa11 Sure.

@lucafei It worked in multi-label segmentation from BCHW to BCHW.

**Comment by CheeseInTrap** [comment]
Hi, I wonder why there is no normalize operation when calculating edt in HausdorffDTLoss, the loss value may be too large
